x-pack/dockerlogbeat/magefile.go

```diff
@@ -47,132 +57,289 @@ func getPluginName() (string, error) {
+	err = sh.RunV("tar", "-xf", dockerExportPath, "-C", filepath.Join(buildDir, "rootfs"))
+	if err != nil {
+		return errors.Wrap(err, "error unpacking exported container")
+	}
+
+	return nil
+}
+
+func cleanDockerArtifacts(containerID string, cli *client.Client) error {
+	fmt.Printf("Removing container %s\n", containerID)
+	err := cli.ContainerRemove(context.Background(), containerID, types.ContainerRemoveOptions{RemoveVolumes: true, Force: true})
+	if err != nil {
+		return errors.Wrap(err, "error removing container")
+	}
+
+	resp, err := cli.ImageRemove(context.Background(), rootImageName, types.ImageRemoveOptions{Force: true})
+	if err != nil {
+		return errors.Wrap(err, "error removing image")
+	}
+	fmt.Printf("Removed image: %#v\n", resp)
+	return nil
 }
 
-// CleanDocker removes working objects and containers
-func CleanDocker() error {
+// Uninstall removes working objects and containers
+func Uninstall() error {
 	name, err := getPluginName()
 	if err != nil {
 		return err
 	}
 
-	sh.RunV("docker", "rm", "-vf", containerName)
-	sh.RunV("docker", "rmi", "rootfsimage")
-	sh.Rm(packageStagingDir)
-	sh.RunV("docker", "plugin", "disable", "-f", name)
-	sh.RunV("docker", "plugin", "rm", "-f", name)
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return errors.Wrap(err, "Error creating docker client")
+	}
+
+	//check to see if we have a plugin we need to remove
+	plugins, err := cli.PluginList(context.Background(), filters.Args{})
+	if err != nil {
+		return errors.Wrap(err, "error getting list of plugins")
+	}
+	oursExists := false
+	for _, plugin := range plugins {
+		if strings.Contains(plugin.Name, logDriverName) {
+			oursExists = true
+		}
+	}
+	if oursExists {
+		err = cli.PluginDisable(context.Background(), name, types.PluginDisableOptions{Force: true})
+		if err != nil {
+			return errors.Wrap(err, "error disabling plugin")
+		}
+		err = cli.PluginRemove(context.Background(), name, types.PluginRemoveOptions{Force: true})
+		if err != nil {
+			return errors.Wrap(err, "error removing plugin")
+		}
+	}
 
 	return nil
 }
 
-// Install installs the beat
+// Install installs the plugin
 func Install() error {
+	mg.Deps(Uninstall)
+	if _, err := os.Stat(filepath.Join(packageStagingDir, "rootfs")); os.IsNotExist(err) {
+		mg.Deps(Build)
+	}
+
 	name, err := getPluginName()
 	if err != nil {
 		return err
 	}
 
-	err = sh.RunV("docker", "plugin", "create", name, packageStagingDir)
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return errors.Wrap(err, "Error creating docker client")
+	}
+
+	archiveOpts := &archive.TarOptions{
+		Compression:  archive.Uncompressed,
+		IncludeFiles: []string{"rootfs", "config.json"},
+	}
+	archive, err := archive.TarWithOptions(buildDir, archiveOpts)
+	if err != nil {
+		return errors.Wrap(err, "error creating archive of work dir")
+	}
+
+	err = cli.PluginCreate(context.Background(), archive, types.PluginCreateOptions{RepoName: name})
```

Comment:
If we don't reuse by now the contexts provided by mage I think that we should use `context.TODO()` instead of all these `context.Background()`.